spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/SparkCachedTableCatalog.java

```diff
@@ -144,23 +149,49 @@ private Pair<Table, Long> load(Identifier ident) throws NoSuchTableException {
       if (snapshotBasedMatcher.matches()) {
         snapshotId = Long.parseLong(snapshotBasedMatcher.group(1));
       }
+
+      Matcher branchBasedMatcher = BRANCH.matcher(meta);
+      if (branchBasedMatcher.matches()) {
+        branch = branchBasedMatcher.group(1);
+      }
+
+      Matcher tagBasedMatcher = BRANCH.matcher(meta);
+      if (tagBasedMatcher.matches()) {
+        tag = tagBasedMatcher.group(1);
+      }
     }
 
     Preconditions.checkArgument(
-        asOfTimestamp == null || snapshotId == null,
-        "Cannot specify both snapshot and timestamp for time travel: %s",
-        ident);
+        Stream.of(snapshotId, asOfTimestamp, branch, tag).filter(Objects::nonNull).count() <= 1,
+        "Can specify at most one of snapshot-id (%s), as-of-timestamp (%s), and snapshot-ref (%s)",
+        snapshotId,
+        asOfTimestamp,
+        branch);
 
     Table table = TABLE_CACHE.get(key);
 
     if (table == null) {
       throw new NoSuchTableException(ident);
     }
 
+    long snapshotIdFromTimeTravel = -1L;
```

Comment:
nit: can use `Optional<Long>` instead of using -1 to represent no snapshot ID

Reply:
Actually, after reading all the if else statements, I am not sure if we really need this, why not just do `return Pair.of(table, table.snapshot(branch).snapshotId())` and `return Pairs.of(table, table.snapshot(tag).snapshotId())` in the branch and tag case?

Reply:
I see, this is a comment from Amogh

> Nit: could we just have each if block match and resolve the snapshot ID, and then just have one return new SparkTable(table, snapshotId, !cacheEnabled) at the end. That'll reduce duplication considering we need to also handle the tag case also.

I actually think the other way, because if we look at this current change, it introduces more changes and variables without the stated benefit of really removing much duplication.

Reply:
Ah yeah, most of the duplication comes from having the different time travel options but ultimately we end up having to have a branching logic anyways to assign the snapshot id. I'm good to revert back to the existing way, we should generally strive to minimize changes as long as it's still readable which it is in the existing case as well. sorry for the confusion there @namrathamyske !